Source/WebCore/html/canvas/CanvasStyle.cpp

```diff
@@ -159,8 +159,10 @@ void CanvasStyle::applyStrokeColor(GraphicsContext& context) const
         [&context] (const Color& color) {
             context.setStrokeColor(color);
         },
-        [&context] (const RefPtr<CanvasGradient>& gradient) {
-            context.setStrokeGradient(gradient->gradient());
+        [&context] (const WeakPtr<CanvasGradient>& gradient) {
+            if (gradient.get())
+                return context.setStrokeGradient(gradient->gradient());
+            ASSERT_NOT_REACHED();
         },
         [&context] (const RefPtr<CanvasPattern>& pattern) {
```

Comment:
Based on your diagnosis, I feel like we should be very suspicious of this `RefPtr<CanvasPattern>`, since that's just another kind of Gradient thing that might be created someplace and create another retain cycle. But perhaps that would make a great second patch.

Reply:
I checked and the CanvasPattern does not contain any references to the context. If that changes in the future I expect the test I added to catch that.